### PR TITLE
zipkin: add key name to prevent bad replacements

### PR DESCRIPTION
**Pull request by douglas-reid**
Commit Message: zipkin: add key name to prevent bad replacements

Additional Description: This PR attempts to fix issues that arise with the current scheme for replacements in the Zipkin serialization that was added in #10400. In particular, this PR is trying to prevent issues that occur when size tags have the same value as the span durations. This was originally reported in https://github.com/istio/istio/issues/24177.

Risk Level: Low
Testing: Unit (with and without changes to validate)
Docs Changes: N/A
Release Notes: N/A

Reviewers: I'm not entirely comfortable with this change, as it feels like adding a hack on top of another hack, but _something_ is needed and this seemed to be the most straightforward way to correct. Hopefully, the original TODO for a better fix will cover this case as well. Please let me know if you would like to pursue a different or clearer path.

Signed-off-by: Douglas Reid <douglas-reid@users.noreply.github.com>

### CODE REVIEW

**Comment by mandarjog** [comment]
@douglas-reid it would be good to check if the open census / zipkin tracer faces the same issue.
There is already a proposal to switch to using that in istio.

**Comment by douglas-reid** [comment]
@mandarjog yup. we need to look at that too. but at the moment, we need a fix we can backport to 1.5.x, etc.

File: source/extensions/tracers/zipkin/span_buffer.cc [hunk 1]
```diff
@@ -137,7 +137,8 @@ JsonV2Serializer::toListOfSpans(const Span& zipkin_span, Util::Replacements& rep
       // us 1.58432429547687e+15. Instead we store it as the string of 1584324295476870 (when it is
       // serialized: "1584324295476870"), and replace it post MessageToJsonString serialization with
       // integer (1584324295476870 without `"`), see: JsonV2Serializer::serialize.
-      (*fields)[SPAN_TIMESTAMP] = Util::uint64Value(annotation.timestamp(), replacements);
+      (*fields)[SPAN_TIMESTAMP] =
+          Util::uint64Value(annotation.timestamp(), "timestamp", replacements);
```
Comment (dio) [inline source/extensions/tracers/zipkin/span_buffer.cc:141]: Can we have this "timestamp" as a constant?

**Comment by dio** [comment]
Yeah, sorry that I've been down for sickness so I couldn't spend time pursuing changes to protobuf upstream instead :(. I think this one seems OK for now.

File: source/extensions/tracers/zipkin/util.h [hunk 2]
```diff
@@ -49,13 +49,16 @@ class Util {
...
+   * replacement to the given replacements vector. The replacement includes the supplied name
+   * as a key, for identification in a JSON stream.
    *
    * @param value unt64_t number that will be represented in string.
+   * @param name std::string that is the key for the value being replaced.
    * @param replacements a container to hold the required replacements when serializing this value.
    * @return ProtobufWkt::Value wrapped uint64_t as a string.
    */
-  static ProtobufWkt::Value uint64Value(uint64_t value, Replacements& replacements);
+  static ProtobufWkt::Value uint64Value(uint64_t value, std::string name,
```
Comment (dio) [inline source/extensions/tracers/zipkin/util.h:60]: when we have the name as constant, probably we can use `absl::string_view` as name here.

File: source/extensions/tracers/zipkin/span_buffer.cc (hunk 1 above)
Comment (douglas-reid) [reply source/extensions/tracers/zipkin/span_buffer.cc]: Yes. Done. Used `SPAN_TIMESTAMP` (and `SPAN_DURATION`).

File: source/extensions/tracers/zipkin/util.h (hunk 2 above)
Comment (douglas-reid) [reply source/extensions/tracers/zipkin/util.h]: Switched.

**Comment by douglas-reid** [comment]
Any updates here?  How can we move this forward?

**Pull request merged**